sql/plan/create_index.go

```diff
@@ -254,13 +255,59 @@ func getColumnsAndPrepareExpressions(
 		})
 
 		if err != nil {
-			return nil, nil, err
+			return nil, nil, nil, err
 		}
 
-		expressions[i] = sql.NewExpressionHash(ex)
+		expressions[i] = ex
+		expressionHashes[i] = sql.NewExpressionHash(ex)
+	}
+
+	return columns, expressions, expressionHashes, nil
+}
+
+type evalKeyValueIter struct {
+	iter sql.IndexKeyValueIter
+	eval func(values []interface{}) ([]interface{}, error)
+}
+
+func (eit *evalKeyValueIter) Next() ([]interface{}, []byte, error) {
+	vals, loc, err := eit.iter.Next()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if eit.eval != nil {
+		vals, err = eit.eval(vals)
+	}
+	return vals, loc, err
+}
+
+func (eit *evalKeyValueIter) Close() error {
+	return eit.iter.Close()
+}
+
+func getIndexKeyValueIter(ctx *sql.Context, table sql.Indexable, columns []string, exprs []sql.Expression) (*evalKeyValueIter, error) {
+	iter, err := table.IndexKeyValueIter(ctx, columns)
+	if err != nil {
+		return nil, err
 	}
 
-	return columns, expressions, nil
+	eit := &evalKeyValueIter{iter: iter}
+	eit.eval = func(values []interface{}) ([]interface{}, error) {
```

Comment:
what do you think about passing all needed stuff to the iterator and move this closure to a method in the `evalKeyValueIter` instead?

Reply:
Generally I'm fine with that (basically this is what I previously did in your `loggingKeyValueIter`). The only concerns I had:
- I don't want/like keep `*sqlContext` in structure (it's more to control execution, imo).
- I have the same feeling as you had that adding all `expressions`, `context` and `iterator` makes the structure too heavy.
- Right now creating `&evalKeyValueIter{iter:iter}` (without `eval` implementation) gives you original iterator, what I found maybe useful in testing.

*BUT* it can be just my taste/perception, so if current structure looks complicated/unreadable in your opinion then I'm totally fine to replace _eval closure_ by more fields.

Reply:
Any concerns proofs that it can be potential readability issue . So I've replaced that closure by straight forward, concrete struct and moved logic to `Next`.